File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        // return listInt(head);
        // return stackSum(head);
        return revInPlace(head);
    }

    int listInt(ListNode* head) {
        ListNode* curr = head;
        vector<int> values;
        
        //iterate over nodes and add ints to values
        while(curr) {
            values.push_back(curr->val);
            curr = curr->next;
        }

        int i = 0; 
        int j = values.size() - 1;
        int maximumSum = 0;

        while (i < j) {
            maximumSum = max(maximumSum, values[i] + values[j]);
            i++;
            j--;
        }

        return maximumSum;
    } //runtime : O(N), space : O(N)

    int stackSum(ListNode* head) {
        ListNode* curr = head;
        stack<int> st;

        while(curr) {
            st.push(curr->val);
            curr = curr->next;
        }

        curr = head;
        int size = st.size(), count = 1;
        int maximumSum = 0;
        while (count <= size) {
            maximumSum = max(maximumSum, curr->val + st.top());
            curr = curr->next;
            st.pop();
            count++;
        }

        return maximumSum;
    } //runtime : O(N), space : O(N)

    int revInPlace(ListNode* head) {
        //get middle of list
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast && fast->next) {
            fast = fast->next->next;
            slow = slow->next;
        }

        //reverse second half of list
        ListNode *nextNode, *prevNode = nullptr;
        while(slow) {
            nextNode = slow->next;
            slow->next = prevNode;
            prevNode = slow;
            slow = nextNode;
        }

        //get max
        int maximumSum = 0;
        ListNode* start = head;
        while(prevNode) {
            maximumSum = max(maximumSum, start->val + prevNode->val);
            prevNode = prevNode->next;
            start = start->next;
        }

        return maximumSum;
    } //runtime O(N), space : O(1)
};

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        if head == None: return False
        
        #get middle
        p1 = p2 = head
        while p2 and p2.next:
            p1 = p1.next
            p2 = p2.next.next
        
        #reverse second half
        prev = None
        curr = p1
        while curr:
            nextTemp = curr.next
            curr.next = prev
            prev = curr 
            curr = nextTemp
        
        #calc max
        maxTwin = -9999
        curTwin = 0
        p1 = prev
        p2 = head
        while p1:
            currTwin = p1.val + p2.val
            maxTwin = max(maxTwin, currTwin)
            p1 = p1.next
            p2 = p2.next
            
        return maxTwin